### ISSUE TITLE
ParsingContext.getCurrentParsedContent() returns null?

### ISSUE THREAD

**Issue by icemanltd** [uniVocity/univocity-parsers#106]
I am using a delimited parser, parsing a pipe delimited file. In my RowProcessor when I attempt to get the currentParsedContent() from ParsingContext I get null. Based on the documentation I was expecting to get the raw data that was parsed to produce the current record.

If my understanding is correct this appears to be a bug? If I am mistaken, is there some way to get the raw input for a given record?

**Comment by jbax** [comment]
Yes that's a bug! Thank you for reporting it. We will get this fixed asap

**Comment by jbax** [comment]
Fixed! I've deployed a new 2.2.1-SNAPSHOT build that includes the required adjustments.

Thank you for using our parsers!

**Comment by icemanltd** [comment]
And thank you for the quick response AND fix. I will update.

**Comment by icemanltd** [comment]
I have an implementation of RowProcessor wrapped in a ConcurrentRowProcessor. Is it valid to call ParsingContext.currentParsedContent() from within my implementation of  RowProcessor.rowProcessed() in this case?
I am either getting null or some other content but not necessarily the content for the current row.

**Comment by jbax** [comment]
Ah that's a very special case because when you call `getCurrentParsedContent()` the parser will grab the start and end of the last marked record from the current state of the input. This will fail miserably when the processing is not sequential. Collecting the relevant content upfront, even though most users will never ask for it, will impact performance greatly.

I'll adjust the the `ConcurrentRowProcessor` to optionally pass on a copy of the `ParsingContext` object at the time a row was parsed and submitted to the internal queue, so that all methods reflect the the state of the parser at the time the rows was parsed.

**Comment by icemanltd** [comment]
Thank you. That is what I thought was going on based on what I was seeing and after reading through some of the code. I am just now exploring the APIs and I am not sure we will end up using the concurrent processor. Very simple use cases so far have shown that the standard processing is as fast if not faster than using the concurrent processor, as you docs suggest.

**Comment by jbax** [comment]
Regardless, I've added the `contextCopyingEnabled` flag to `ConcurrentRowProcessor` which will make it "save" all variable elements of the context of each parsed record so the behavior becomes deterministic.

The latest changes are now available on the latest `2.2.1-SNAPSHOT` build 

Thanks!

**Comment by icemanltd** [comment]
I think there is still an issue. When calling, ParsingContext.currentParsedContent() I get a StringIndexOutOfBoundsException, regardless of whether I am calling from the concurrent row processor or whether I simply iterating via the parser's parseNext() method.
When calling ParsingContext.currentParsedContext() while iterating:
```
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.<init>(String.java:192)
	at com.univocity.parsers.common.input.AbstractCharInputReader.currentParsedContent(AbstractCharInputReader.java:320)
	at com.univocity.parsers.common.DefaultParsingContext.currentParsedContent(DefaultParsingContext.java:67)
```
When calling ParsingContext.currentParsedContent() from within the concurrent row processor:
```
Exception in thread "main" java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.lang.String.<init>(String.java:192)
	at com.univocity.parsers.common.input.AbstractCharInputReader.currentParsedContent(AbstractCharInputReader.java:320)
	at com.univocity.parsers.common.DefaultParsingContext.currentParsedContent(DefaultParsingContext.java:67)
	at com.univocity.parsers.common.ParsingContextSnapshot.<init>(ParsingContextSnapshot.java:49)
	at com.univocity.parsers.common.processor.ConcurrentRowProcessor.copyContext(ConcurrentRowProcessor.java:53)
	at com.univocity.parsers.common.processor.ConcurrentRowProcessor.copyContext(ConcurrentRowProcessor.java:29)
	at com.univocity.parsers.common.processor.core.AbstractConcurrentProcessor.grabContext(AbstractConcurrentProcessor.java:220)
	at com.univocity.parsers.common.processor.core.AbstractConcurrentProcessor.processEnded(AbstractConcurrentProcessor.java:214)
	at com.univocity.parsers.common.AbstractParser.stopParsing(AbstractParser.java:394)
```

I suspect it may be happening when the buffer is being reloaded. I apologize that I have not had more time to try and track it further.

**Comment by jbax** [comment]
Thanks for reporting this one! I've fixed it and released another 2.2.1-SNAPSHOT build
